### PR TITLE
[be2-scala] Parser implementation with new Json protocol

### LINKED ISSUE #69: Update Parser/Messages to new protocol version
Update Parser and messages to the new-new protocol : 

- [x]  define new `JsonMessages` (`JsonParser` output) and types
- [x]  define client queries JSON serialization (low-level communication) ⚠️ *omitted "publish a message on a channel"*
- [x]  define client queries JSON serialization (high-level communication)
- [x]  define server answers JSON encoding *~(50% done)~*
- [x]  define error codes return (send error codes instead of exceptions back to client) *(50% done)*
- [x]  write unit tests for all `JsonMessages` (see `JsonMessageParserTest`)
- [x]  update `PubSub`, `DBActor` and `Server` dependencies (huge thanks to @SFulpius)

### PR DESCRIPTION
**Pull request by RaulinN**
Closes #69

### CODE REVIEW

**Comment by SFulpius** [comment]
I added the LAO creation messages, so we can test the code. I will adapt the test so they compile and it should be good.

**Review by SFulpius — CHANGES_REQUESTED** [review]
Great job! Parsing three levels of Json was not an easy task!
There are a few minor mistakes, please look at my comments

File: src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    val InvalidResource: JsonAnswerServer.ErrorCodes.Value = Value(-2, "invalid resource")
+    // resource already exists (e.g. lao already exists, channel already exists, etc.)
+    val AlreadyExists: JsonAnswerServer.ErrorCodes.Value = Value(-3, "resource already exists")
+    // request data is invalid (e.g. message is invalid)
+    val InvalidData: JsonAnswerServer.ErrorCodes.Value = Value(-4, "request data is invalid")
+    // access denied (e.g. subscribing to a “restricted” channel)
+    val AccessDenied: JsonAnswerServer.ErrorCodes.Value = Value(-5, "access denied")
+  }
+
+  def sendAnswer(answer: String): Unit = {
```
[the remark this answers is not in the document]
Comment (RaulinN) [reply src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala:27]: I am still not sure how to handle errors. This function only prints them to the terminal as a placeholder

File: src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import ch.epfl.pop.json.JsonAnswerServer.ErrorCodes.ErrorCodes
+import ch.epfl.pop.json.JsonMessages.{AnswerErrorMessageServer, AnswerResultArrayMessageServer, AnswerResultIntMessageServer}
+
+
+
+object JsonAnswerServer {
+
+  object ErrorCodes extends Enumeration {
+    type ErrorCodes = Value
```
[the remark this answers is not in the document]
Comment (RaulinN) [reply src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala:11]: `Value` represents the type of the enumeration. It comes from the [`Enumeration` abstract class](https://www.scala-lang.org/api/current/scala/Enumeration.html)

**Comment by RaulinN** [comment]
> Actually `PropagateMessageServer` is a server message, so I don't see why it would make a problem. Also, I dont understand your problem with `id`. For me, it should work without problem. Is it a problem with the parser?

Interestingly enough, this creates a `java.lang.ExceptionInInitializerError`. I have not found why this doesn't work. Moreover there's one lonely Stackoverflow post only mentioning this exception online which speaks about implicit val orders... But I have triple-checked ten times the order 🤔 . I will look into it again soon to see if there's really no way to make it work

File: src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (SFulpius) [reply src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala:11]: It does not compile on my computer. Does it works on yours?

[the remark this answers is not in the document]
Comment (SFulpius) [reply src/main/scala/ch/epfl/pop/json/JsonAnswerServer.scala:11]: Looks like it was a bug in my code. It compiles now.

**Review by SFulpius — APPROVED** [review]

**Comment by SFulpius** [comment]
> > Actually `PropagateMessageServer` is a server message, so I don't see why it would make a problem. Also, I dont understand your problem with `id`. For me, it should work without problem. Is it a problem with the parser?
> 
> Interestingly enough, this creates a `java.lang.ExceptionInInitializerError`. I have not found why this doesn't work. Moreover there's one lonely Stackoverflow post only mentioning this exception online which speaks about implicit val orders... But I have triple-checked ten times the order thinking . I will look into it again soon to see if there's really no way to make it work

When do you get this exception? I tried with this code 
```
sealed class JsonMessagePublishClient(
                                       val params: MessageParameters,
                                       val id: Int,
                                       val method: Methods,
                                       val jsonrpc: String = JSON_RPC_VERSION
                                     ) extends JsonMessagePubSubClient
```

and I get no error running the tests.

**Comment by SFulpius** [comment]
We will not add further features to this PR.
@lhmerino or @haoqianzhang if you have some time before tomorrow's meeting to review this PR, it would be appreciated :smile:

**Review by lhmerino — APPROVED** [review]
Looks good! The code is well documented. There are some TODO comments which I assume will be addressed in a future pull request.

**Comment by SFulpius** [comment]
Thank you for your review!
You're right, TODO are for future work, as the PR was already long enough.

**Pull request merged**